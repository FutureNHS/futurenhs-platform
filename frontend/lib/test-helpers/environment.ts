export const withEnvVars = (vars: { [name: string]: string }) => {
  beforeAll(() => {
    Object.assign(process.env, vars);
  });

  afterAll(() => {
    for (const key of Object.keys(vars)) {
      delete process.env[key];
    }
  });
};
