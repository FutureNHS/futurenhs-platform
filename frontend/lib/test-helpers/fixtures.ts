export const kratosRequestBase = {
  active: "password",
  expiresAt: new Date(),
  id: "598a45d2-e107-41ba-885a-c5c39e4a26d5",
  issuedAt: new Date(),
  messages: undefined,
  // I don't think we ever use this.
  requestUrl: "http://localhost:4455/self-service/browser/flows/settings",
};
