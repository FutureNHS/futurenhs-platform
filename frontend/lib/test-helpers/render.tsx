import React from "react";

import { render as defaultRender, RenderResult } from "@testing-library/react";
import { RouterContext } from "next/dist/next-server/lib/router-context";
import { NextRouter } from "next/router";
import { ThemeProvider } from "styled-components";

import theme from "../../lib/fixtures/theme.json";

export * from "@testing-library/react";

// --------------------------------------------------
// Override the default test render with our own, which includes a router and
// theme mock.
//
// You can override the router mock like this:
//
// const { baseElement } = render(<MyComponent />, {
//   router: { pathname: '/my-custom-pathname' },
// });
// --------------------------------------------------
type DefaultParams = Parameters<typeof defaultRender>;
type RenderUI = DefaultParams[0];
type RenderOptions = DefaultParams[1] & { router?: Partial<NextRouter> };

export function render(
  ui: RenderUI,
  { wrapper, router, ...options }: RenderOptions = {}
): RenderResult {
  if (!wrapper) {
    const Wrapper = ({ children }: { children?: React.ReactNode }) => (
      <RouterContext.Provider value={{ ...mockRouter, ...router }}>
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
      </RouterContext.Provider>
    );
    Wrapper.displayName = "Wrapper";
    wrapper = Wrapper;
  }

  return defaultRender(ui, { wrapper, ...options });
}

const mockRouter: NextRouter = {
  basePath: "",
  pathname: "/",
  route: "/",
  asPath: "/",
  query: {},
  push: jest.fn(),
  replace: jest.fn(),
  reload: jest.fn(),
  back: jest.fn(),
  prefetch: jest.fn(),
  beforePopState: jest.fn(),
  events: {
    on: jest.fn(),
    off: jest.fn(),
    emit: jest.fn(),
  },
  isFallback: false,
};
