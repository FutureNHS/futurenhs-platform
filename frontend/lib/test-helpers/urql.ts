import { DocumentNode } from "graphql";
import { makeResult, Client, Operation, createClient } from "urql";
import { fromValue } from "wonka";

import { Query } from "../generated/graphql";

export type MockUrqlClient = (
  queries: Array<[DocumentNode, Partial<Query>]>
) => Client;

export const mockUrqlClient: MockUrqlClient = (queries) => {
  const client = createClient({ url: "http://localhost", exchanges: [] });
  return Object.assign(client, {
    executeQuery: jest.fn(({ query }) => {
      const operation: Operation = {
        context: {} as any,
        key: 1,
        operationName: "query",
        query,
      };
      const data = queries.find((q) => q[0] === query);
      return fromValue(
        data
          ? makeResult(operation, { data: data[1] })
          : makeResult(operation, {
              error: "query not found",
            })
      );
    }),
  });
};
