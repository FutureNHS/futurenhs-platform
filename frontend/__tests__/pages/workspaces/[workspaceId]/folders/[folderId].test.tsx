import React from "react";

import { ThemeProvider } from "styled-components";
import { fromValue, never } from "wonka";

import theme from "../../../../../lib/fixtures/theme.json";
import { render } from "../../../../../lib/test-helpers/render";
import FolderHomepage from "../../../../../pages/workspaces/[workspaceId]/folders/[folderId]";

describe(FolderHomepage, () => {
  const responseState = {
    executeQuery: jest.fn(({ query }) => {
      if (
        query.definitions.find((d: any) => d.name.value == "GetWorkspaceByID")
      ) {
        return fromValue({
          data: {
            workspace: {
              id: "1",
              title: "hospital",
            },
          },
        });
      }
      if (query.definitions.find((d: any) => d.name.value == "GetFolderById")) {
        return fromValue({
          data: {
            folder: {
              id: "f1",
              title: "folder 1",
              description: "first folder",
              workspace: "1",
            },
          },
        });
      }
      if (
        query.definitions.find((d: any) => d.name.value == "FoldersByWorkspace")
      ) {
        return fromValue({
          data: {
            foldersByWorkspace: [
              {
                id: "f1",
                title: "folder 1",
                description: "first folder",
                workspace: "1",
              },
            ],
          },
        });
      }
    }),
    executeMutation: jest.fn(() => never),
    executeSubscription: jest.fn(() => never),
  };

  test("renders matching snapshot", () => {
    const container = render(
      <ThemeProvider theme={theme}>
        <FolderHomepage urqlClient={responseState} />
      </ThemeProvider>,
      {
        router: {
          query: { workspaceId: "1", folderId: "f1" },
        },
      }
    );
    expect(container.asFragment()).toMatchSnapshot();
  });
});
