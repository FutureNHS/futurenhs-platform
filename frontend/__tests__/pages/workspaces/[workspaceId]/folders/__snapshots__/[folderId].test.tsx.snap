// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderHomepage renders matching snapshot 1`] = `
<DocumentFragment>
  .c1 {
  background-color: rgb(255,255,255);
  padding: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  display: block;
  height: 40px;
}

.c13 {
  display: inline-block;
  position: relative;
}

.c15 {
  opacity: 0;
  position: absolute;
  middle: 50%;
  left: 100%;
  margin-left: 13px;
  z-index: 1000;
  border-radius: 4px;
  background-color: rgb(76,98,114);
  padding: 5px;
  text-align: center;
  font-weight: 400;
  font-size: 18px;
  color: rgb(255,255,255);
}

.c12:hover .c14 {
  opacity: 1;
  -webkit-transition: opacity 0s linear 0.5s;
  transition: opacity 0s linear 0.5s;
}

.c15::after {
  position: absolute;
  bottom: 33%;
  left: -10%;
  margin-left: -2px;
  border-width: 5px;
  border-style: solid;
  border-color: transparent rgb(76,98,114) transparent transparent;
  content: " ";
}

.c19 {
  display: inline-block;
  width: 24px;
  height: 24px;
}

.c11 {
  border: 0;
  padding: 0;
  width: 24px;
  height: 24px;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
}

.c18 {
  margin-bottom: 23px;
}

.c20 {
  margin-top: 0px;
  margin-bottom: 25px;
  color: rgb(76,98,114);
  border-bottom: 1px solid rgb(76,98,114);
}

.c9 {
  list-style-type: none;
  margin-bottom: 12px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c10 {
  background: rgb(255,235,59);
  border-radius: 4px;
  width: 227px;
  margin-right: 4px;
  font-weight: 700;
}

.c10:hover {
  background-color: rgb(216,221,224);
}

.c10 a {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-left: 8px;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c10 a:focus {
  box-shadow: none;
}

.c10 img {
  padding-top: 2px;
  -webkit-align-self: baseline;
  -ms-flex-item-align: baseline;
  align-self: baseline;
  width: 24px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c10 div {
  padding-left: 4px;
  padding-right: 4px;
  color: rgb(33,43,50);
}

.c7 {
  padding-top: 20px;
}

.c7 div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: start;
  -webkit-box-align: start;
  -ms-flex-align: start;
  align-items: start;
}

.c7 button {
  border: none;
  background: inherit;
}

.c7 h4 {
  color: rgb(0,94,184);
}

.c4 {
  padding-top: 24px;
  padding-right: 52px;
  padding-left: 20px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  box-sizing: border-box;
  width: 270px;
}

.c5 {
  padding-bottom: 20px;
  border-bottom: 1px solid rgb(76,98,114);
}

.c6 {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
}

.c6:hover {
  color: inherit;
}

.c6:visited {
  color: inherit;
}

.c8 {
  padding-left: 0px;
}

.c0 {
  max-width: 1200px;
  margin: 0px auto;
}

.c16 {
  -webkit-box-flex: 3;
  -webkit-flex-grow: 3;
  -ms-flex-positive: 3;
  flex-grow: 3;
  min-height: 100vh;
  padding-top: 24px;
  padding-left: 10%;
  padding-right: 10%;
  background-color: rgb(255,255,255);
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media (max-width:641px) {
  .c2 {
    height: 32px;
  }
}

@media (max-width:320px) {
  .c2 {
    height: 28px;
  }
}

@media (min-width:641px) {
  .c4 {
    width: 300px;
  }
}

@media (min-width:990px) {
  .c4 {
    width: 360px;
  }
}

<div
    class="c0"
  >
    <header
      class="c1"
    >
      <img
        alt="FutureNHS logo"
        class="c2"
        src="/public/FutureNHS.svg"
      />
      <a
        href="https://www.nhs.co.uk"
      >
        <img
          alt="https://www.nhs.co.uk"
          class="c2"
          src="/public/NHS.svg"
        />
      </a>
    </header>
    <div
      class="c3"
    >
      <nav
        class="c4"
      >
        <header
          class="c5"
        >
          <a
            class="c6"
            href="/workspaces/1"
          >
            <h3>
              hospital
            </h3>
          </a>
          <a
            href="/workspaces/1"
          >
            About this workspace
          </a>
        </header>
        <section
          class="c7"
        >
          <div>
            <h4>
              Folders
            </h4>
            <button>
              <img
                alt="Hide Folders"
                src="/public/chevronOpen.svg"
              />
            </button>
          </div>
          <ul
            class="c8"
          >
            <li
              class="c9"
            >
              <div
                class="c10"
              >
                <a
                  href="/workspaces/1/folders/f1"
                >
                  <img
                    src="/public/folderOpen.svg"
                  />
                  <div>
                    folder 1
                  </div>
                </a>
              </div>
              <button
                class="c11"
              >
                <div
                  class="c12 c13"
                >
                  <span
                    class="c14 c15"
                  >
                    Options
                  </span>
                </div>
              </button>
            </li>
          </ul>
        </section>
      </nav>
      <section
        class="c16"
      >
        <div
          class="c17"
        >
          <h1
            aria-live="polite"
            class="c18"
          >
            folder 1
          </h1>
          <button
            class="c11"
          >
            <div
              class="c12 c13"
            >
              <svg
                class="c19"
                fill="none"
                height="1em"
                viewBox="0 0 24 24"
                width="1em"
              >
                <rect
                  fill="rgb(216, 221, 224)"
                  height="24"
                  rx="4"
                  width="24"
                />
                <path
                  clip-rule="evenodd"
                  d="M17 12a2 2 0 104 0 2 2 0 00-4 0zm-5 2a2 2 0 110-4 2 2 0 010 4zm-7 0a2 2 0 110-4 2 2 0 010 4z"
                  fill="rgb(33, 43, 50)"
                  fill-rule="evenodd"
                />
              </svg>
              <span
                class="c14 c15"
              >
                Options
              </span>
            </div>
          </button>
        </div>
        <hr
          class="c20"
        />
        <p>
          first folder
        </p>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
